File: 12. Integer to Roman.py
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        :ref: http://www.cnblogs.com/dosxp/archive/2008/08/13/1266781.html
        """
        hrom = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"]
        trom = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"]
        rom = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"]
        intdex = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
        hdic = dict(zip(intdex, hrom))   
        tdic = dict(zip(intdex, trom))
        dic = dict(zip(intdex, rom))
        
        rostr = ""
        k = num // 1000
        num = num % 1000
    
        h = num // 100
        num = num % 100
        
        t = num // 10
        num = num % 10
        
        rostr = "M"*k + hdic[h] + tdic[t] + dic[num]
        return rostr
        

File: 12. Integer to Roman.cpp
class Solution {
public:
    string intToRoman(int num) {
        char* rom3[4] = {"", "M", "MM", "MMM"};
        char* rom2[10] = {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"}; //0XX - 9XX
        char* rom1[10] = {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"}; //0X - 9X
        char* rom0[10] = {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"}; //0 - 9
        int d[4];
        d[0] = num % 10;
        d[1] = num % 100 / 10;
        d[2] = num % 1000 / 100;
        d[3] = num / 1000;
        string s;
        s += rom3[d[3]];
        s += rom2[d[2]];
        s += rom1[d[1]];
        s += rom0[d[0]];
        return s;
    }
};
